File: apps/web/src/types/delta.ts
// TypeScript mirror of uh-core::schema::delta

import type { MatchVerdict, StepId, StepIntent, PlanId } from "./plan";

export type DeltaTarget =
  | { plan: PlanId }
  | { step: { plan_id: PlanId; step_id: StepId } };

export interface StepSpec {
  title: string;
  intent: StepIntent;
  what: string;
  how: string | null;
  risk: string | null;
  backout: string | null;
}

export type DeltaAction =
  | { type: "approve_plan" }
  | { type: "reject_plan"; reason: string }
  | { type: "set_goal"; text: string }
  | { type: "add_criterion"; text: string }
  | { type: "remove_criterion"; id: string }
  | { type: "add_assumption"; text: string; critical: boolean }
  | { type: "remove_assumption"; id: string }
  | { type: "add_unknown"; text: string; blocking: boolean }
  | { type: "remove_unknown"; id: string }
  | { type: "start_step" }
  | {
      type: "complete_step";
      actual: string;
      matches_plan: MatchVerdict;
      deviation: string | null;
    }
  | { type: "fail_step"; error: string }
  | { type: "skip_step"; reason: string }
  | { type: "approve_step" }
  | { type: "reject_step"; reason: string }
  | { type: "add_step"; after_step: StepId | null; spec: StepSpec }
  | { type: "remove_step" }
  | { type: "set_step_title"; text: string }
  | { type: "set_step_what"; text: string }
  | { type: "set_step_how"; text: string | null }
  | { type: "set_step_risk"; text: string | null }
  | { type: "set_step_backout"; text: string | null }
  | { type: "set_step_intent"; intent: StepIntent }
  | { type: "reorder_steps"; order: StepId[] }
  | { type: "add_comment"; text: string }
  | { type: "ask_question"; question: string }
  | { type: "pause_plan" }
  | { type: "resume_plan" }
  | { type: "abort_plan"; reason: string };

export interface Delta {
  id: string;
  timestamp: string;
  target: DeltaTarget;
  action: DeltaAction;
  comment: string | null;
}

File: crates/uh-core/src/schema/delta.rs
//! Delta types — user interactions with Plans and Steps.
//!
//! Every change a user makes to a Plan goes through a `Delta`. This
//! makes the audit trail complete and lets the LLM react to minimal,
//! structured inputs instead of re-reading the whole plan each turn.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::ids::{PlanId, StepId};
use crate::schema::plan::{MatchVerdict, StepIntent};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Delta {
    pub id: Uuid,
    pub timestamp: DateTime<Utc>,
    pub target: DeltaTarget,
    pub action: DeltaAction,
    /// Optional user comment attached to this delta.
    pub comment: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum DeltaTarget {
    Plan(PlanId),
    Step { plan_id: PlanId, step_id: StepId },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum DeltaAction {
    // ─── Plan-level ─────────────────────────────────────────────
    ApprovePlan,
    RejectPlan {
        reason: String,
    },
    SetGoal {
        text: String,
    },
    AddCriterion {
        text: String,
    },
    RemoveCriterion {
        id: Uuid,
    },
    AddAssumption {
        text: String,
        critical: bool,
    },
    RemoveAssumption {
        id: Uuid,
    },
    AddUnknown {
        text: String,
        blocking: bool,
    },
    RemoveUnknown {
        id: Uuid,
    },

    // ─── Step lifecycle ─────────────────────────────────────────
    StartStep,
    CompleteStep {
        actual: String,
        matches_plan: MatchVerdict,
        deviation: Option<String>,
    },
    FailStep {
        error: String,
    },
    SkipStep {
        reason: String,
    },
    ApproveStep,
    RejectStep {
        reason: String,
    },

    // ─── Step editing ───────────────────────────────────────────
    AddStep {
        after_step: Option<StepId>,
        spec: StepSpec,
    },
    RemoveStep,
    SetStepTitle {
        text: String,
    },
    SetStepWhat {
        text: String,
    },
    SetStepHow {
        text: Option<String>,
    },
    SetStepRisk {
        text: Option<String>,
    },
    SetStepBackout {
        text: Option<String>,
    },
    SetStepIntent {
        intent: StepIntent,
    },
    ReorderSteps {
        order: Vec<StepId>,
    },

    // ─── Free-form ──────────────────────────────────────────────
    AddComment {
        text: String,
    },
    AskQuestion {
        question: String,
    },

    // ─── Flow control ───────────────────────────────────────────
    PausePlan,
    ResumePlan,
    AbortPlan {
        reason: String,
    },
}

/// Template used when adding a new step. Doesn't carry an ID —
/// the system assigns one when the step is created.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct StepSpec {
    pub title: String,
    pub intent: StepIntent,
    pub what: String,
    pub how: Option<String>,
    pub risk: Option<String>,
    pub backout: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn delta_action_is_tagged() {
        let action = DeltaAction::SetGoal {
            text: "new goal".into(),
        };
        let json = serde_json::to_string(&action).unwrap();
        assert!(json.contains("\"type\":\"set_goal\""));
    }
}
